also have a panic in Janky; https://jenkins.dockerproject.org/job/Docker-PRs/30333/console

```
23:14:23 PASS: docker_cli_run_test.go:457: DockerSuite.TestRunVolumesFromInReadonlyModeFails	0.993s
23:19:23 --- FAIL: Test (1473.52s)
23:19:23 panic: DockerSuite.TestRunVolumesFromRestartAfterRemoved test timed out after 5m0s [recovered]
23:19:23 	panic: DockerSuite.TestRunVolumesFromRestartAfterRemoved test timed out after 5m0s
23:19:23 
23:19:23 goroutine 19 [running]:
23:19:23 panic(0xd85d20, 0xc820294080)
23:19:23 	/usr/local/go/src/runtime/panic.go:481 +0x3e6
23:19:23 testing.tRunner.func1(0xc820350360)
23:19:23 	/usr/local/go/src/testing/testing.go:467 +0x192
23:19:23 panic(0xd85d20, 0xc820294080)
23:19:23 	/usr/local/go/src/runtime/panic.go:443 +0x4e9
23:19:23 github.com/go-check/check.(*suiteRunner).runTest(0xc820350480, 0xc8203de2a0, 0xc820441d10)
23:19:23 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:828 +0x302
23:19:23 github.com/go-check/check.(*suiteRunner).run(0xc820350480, 0x16f61a0)
23:19:23 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:618 +0x2a0
23:19:23 github.com/go-check/check.Run(0xfb8a80, 0x16f61a0, 0xc820364460, 0x2)
23:19:23 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/run.go:100 +0x43
23:19:23 github.com/go-check/check.RunAll(0xc820364460, 0x0)
23:19:23 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/run.go:92 +0xc8
23:19:23 github.com/go-check/check.TestingT(0xc820350360)
23:19:23 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/run.go:80 +0x337
23:19:23 github.com/docker/docker/integration-cli.Test(0xc820350360)
23:19:23 	/go/src/github.com/docker/docker/integration-cli/check_test.go:25 +0x10f
23:19:23 testing.tRunner(0xc820350360, 0x16af760)
23:19:23 	/usr/local/go/src/testing/testing.go:473 +0x98
23:19:23 created by testing.RunTests
23:19:23 	/usr/local/go/src/testing/testing.go:582 +0x892
23:19:23 
23:19:23 goroutine 1 [chan receive, 24 minutes]:
23:19:23 testing.RunTests(0x11a6f30, 0x16af760, 0x1, 0x1, 0xc8203ae901)
23:19:23 	/usr/local/go/src/testing/testing.go:583 +0x8d2
23:19:23 testing.(*M).Run(0xc820034f08, 0x16d78a0)
23:19:23 	/usr/local/go/src/testing/testing.go:515 +0x81
23:19:23 main.main()
23:19:23 	github.com/docker/docker/integration-cli/_test/_testmain.go:132 +0x253
23:19:23 
23:19:23 goroutine 17 [syscall, 24 minutes, locked to thread]:
23:19:23 runtime.goexit()
23:19:23 	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1
23:19:23 
23:19:23 goroutine 5 [syscall, 24 minutes]:
23:19:23 os/signal.signal_recv(0x0)
23:19:23 	/usr/local/go/src/runtime/sigqueue.go:116 +0x132
23:19:23 os/signal.loop()
23:19:23 	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
23:19:23 created by os/signal.init.1
23:19:23 	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
23:19:24 
23:19:24 goroutine 20 [select, 5 minutes]:
23:19:24 github.com/go-check/check.(*resultTracker)._loopRoutine(0xc8203503f0)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:461 +0x396
23:19:24 created by github.com/go-check/check.(*resultTracker).start
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:441 +0x35
23:19:24 
23:19:24 goroutine 9039 [syscall, 5 minutes]:
23:19:24 syscall.Syscall6(0x3d, 0x2f4f, 0xc82393772c, 0x0, 0xc820b063f0, 0x0, 0x0, 0xc8237edec0, 0x7fb1c170f5b8, 0xe51da0)
23:19:24 	/usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
23:19:24 syscall.wait4(0x2f4f, 0xc82393772c, 0x0, 0xc820b063f0, 0x90, 0x0, 0x0)
23:19:24 	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:172 +0x7f
23:19:24 syscall.Wait4(0x2f4f, 0xc823937774, 0x0, 0xc820b063f0, 0xc8234eb538, 0x0, 0x0)
23:19:24 	/usr/local/go/src/syscall/syscall_linux.go:256 +0x55
23:19:24 os.(*Process).wait(0xc823991f40, 0xe, 0x0, 0x0)
23:19:24 	/usr/local/go/src/os/exec_unix.go:22 +0x105
23:19:24 os.(*Process).Wait(0xc823991f40, 0x0, 0x0, 0x0)
23:19:24 	/usr/local/go/src/os/doc.go:49 +0x2d
23:19:24 os/exec.(*Cmd).Wait(0xc820b02a00, 0x0, 0x0)
23:19:24 	/usr/local/go/src/os/exec/exec.go:396 +0x211
23:19:24 os/exec.(*Cmd).Run(0xc820b02a00, 0x0, 0x0)
23:19:24 	/usr/local/go/src/os/exec/exec.go:262 +0x64
23:19:24 github.com/docker/docker/pkg/integration.RunCommand(0xc820b02a00, 0xc822247d00, 0x0, 0x0)
23:19:24 	/go/src/github.com/docker/docker/pkg/integration/utils.go:161 +0x2f
23:19:24 github.com/docker/docker/integration-cli.runCommand(0xc820b02a00, 0x60, 0x0, 0x0)
23:19:24 	github.com/docker/docker/integration-cli/_test/_obj_test/utils.go:59 +0x39
23:19:24 github.com/docker/docker/integration-cli.deleteContainer(0xc820295810, 0xc, 0x0, 0x0)
23:19:24 	github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:295 +0x246
23:19:24 github.com/docker/docker/integration-cli.deleteAllContainers(0x0, 0x0)
23:19:24 	github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:331 +0x1a9
23:19:24 github.com/docker/docker/integration-cli.(*DockerSuite).TearDownTest(0x16f61a0, 0xc8236503c0)
23:19:24 	/go/src/github.com/docker/docker/integration-cli/check_test.go:37 +0x1d
23:19:24 reflect.Value.call(0xfb8a80, 0x16f61a0, 0xa13, 0xfc7468, 0x4, 0xc823937f10, 0x1, 0x1, 0x0, 0x0, ...)
23:19:24 	/usr/local/go/src/reflect/value.go:435 +0x120d
23:19:24 reflect.Value.Call(0xfb8a80, 0x16f61a0, 0xa13, 0xc823937f10, 0x1, 0x1, 0x0, 0x0, 0x0)
23:19:24 	/usr/local/go/src/reflect/value.go:303 +0xb1
23:19:24 github.com/go-check/check.(*suiteRunner).runFixture.func1(0xc8236503c0)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:732 +0x163
23:19:24 github.com/go-check/check.(*suiteRunner).forkCall.func1(0xc820350480, 0xc8236503c0, 0x11a6850)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:669 +0x6f
23:19:24 created by github.com/go-check/check.(*suiteRunner).forkCall
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:670 +0x299
23:19:24 
23:19:24 goroutine 5723 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).readLoop(0xc82005cb60)
23:19:24 	/usr/local/go/src/net/http/transport.go:1185 +0xd52
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:860 +0x10a6
23:19:24 
23:19:24 goroutine 5759 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).readLoop(0xc82005c9c0)
23:19:24 	/usr/local/go/src/net/http/transport.go:1185 +0xd52
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:860 +0x10a6
23:19:24 
23:19:24 goroutine 3673 [select, 15 minutes]:
23:19:24 net/http.(*persistConn).readLoop(0xc8203669c0)
23:19:24 	/usr/local/go/src/net/http/transport.go:1185 +0xd52
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:860 +0x10a6
23:19:24 
23:19:24 goroutine 3674 [select, 15 minutes]:
23:19:24 net/http.(*persistConn).writeLoop(0xc8203669c0)
23:19:24 	/usr/local/go/src/net/http/transport.go:1280 +0x472
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:861 +0x10cb
23:19:24 
23:19:24 goroutine 5822 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).writeLoop(0xc820123c70)
23:19:24 	/usr/local/go/src/net/http/transport.go:1280 +0x472
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:861 +0x10cb
23:19:24 
23:19:24 goroutine 5744 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).writeLoop(0xc82005d790)
23:19:24 	/usr/local/go/src/net/http/transport.go:1280 +0x472
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:861 +0x10cb
23:19:24 
23:19:24 goroutine 5743 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).readLoop(0xc82005d790)
23:19:24 	/usr/local/go/src/net/http/transport.go:1185 +0xd52
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:860 +0x10a6
23:19:24 
23:19:24 goroutine 5821 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).readLoop(0xc820123c70)
23:19:24 	/usr/local/go/src/net/http/transport.go:1185 +0xd52
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:860 +0x10a6
23:19:24 
23:19:24 goroutine 5760 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).writeLoop(0xc82005c9c0)
23:19:24 	/usr/local/go/src/net/http/transport.go:1280 +0x472
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:861 +0x10cb
23:19:24 
23:19:24 goroutine 6374 [semacquire, 11 minutes]:
23:19:24 sync.runtime_Syncsemacquire(0xc8200be978)
23:19:24 	/usr/local/go/src/runtime/sema.go:241 +0x201
23:19:24 sync.(*Cond).Wait(0xc8200be968)
23:19:24 	/usr/local/go/src/sync/cond.go:63 +0x9b
23:19:24 io.(*pipe).write(0xc8200be900, 0xc823642000, 0x6, 0x8000, 0x0, 0x0, 0x0)
23:19:24 	/usr/local/go/src/io/pipe.go:94 +0x23a
23:19:24 io.(*PipeWriter).Write(0xc820026fe8, 0xc823642000, 0x6, 0x8000, 0x6, 0x0, 0x0)
23:19:24 	/usr/local/go/src/io/pipe.go:161 +0x50
23:19:24 io.copyBuffer(0x7fb1c1712400, 0xc820026fe8, 0x7fb1c1705540, 0xc820027000, 0xc823642000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
23:19:24 	/usr/local/go/src/io/io.go:382 +0x2c9
23:19:24 io.Copy(0x7fb1c1712400, 0xc820026fe8, 0x7fb1c1705540, 0xc820027000, 0x5, 0x0, 0x0)
23:19:24 	/usr/local/go/src/io/io.go:350 +0x64
23:19:24 os/exec.(*Cmd).writerDescriptor.func1(0x0, 0x0)
23:19:24 	/usr/local/go/src/os/exec/exec.go:236 +0x8b
23:19:24 os/exec.(*Cmd).Start.func1(0xc8235f77c0, 0xc8239aa800)
23:19:24 	/usr/local/go/src/os/exec/exec.go:344 +0x1d
23:19:24 created by os/exec.(*Cmd).Start
23:19:24 	/usr/local/go/src/os/exec/exec.go:345 +0x967
23:19:24 
23:19:24 goroutine 9036 [select, 5 minutes]:
23:19:24 github.com/go-check/check.(*suiteRunner).runFunc(0xc820350480, 0xc8203565b0, 0x0, 0xc822247880, 0x31, 0x0, 0x11a6850, 0x42dfae)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:681 +0x39a
23:19:24 github.com/go-check/check.(*suiteRunner).runFixture(0xc820350480, 0xc8203565b0, 0xc822247880, 0x31, 0x0, 0xc82031bf10)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:728 +0x65
23:19:24 github.com/go-check/check.(*suiteRunner).runFixtureWithPanic(0xc820350480, 0xc8203565b0, 0xc822247880, 0x31, 0x0, 0xc823939ddf, 0x16)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:746 +0x72
23:19:24 github.com/go-check/check.(*suiteRunner).forkTest.func1(0xc8236502d0)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:784 +0x48a
23:19:24 github.com/go-check/check.(*suiteRunner).forkCall.func1(0xc820350480, 0xc8236502d0, 0xc8239912e0)
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:669 +0x6f
23:19:24 created by github.com/go-check/check.(*suiteRunner).forkCall
23:19:24 	/go/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:670 +0x299
23:19:24 
23:19:24 goroutine 5724 [select, 12 minutes]:
23:19:24 net/http.(*persistConn).writeLoop(0xc82005cb60)
23:19:24 	/usr/local/go/src/net/http/transport.go:1280 +0x472
23:19:24 created by net/http.(*Transport).dialConn
23:19:24 	/usr/local/go/src/net/http/transport.go:861 +0x10cb
23:19:24 ---> Making bundle: .integration-daemon-stop (in bundles/1.12.0-dev/test-integration-cli)
```

--- a/integration-cli/docker_cli_plugins_test.go
+++ b/integration-cli/docker_cli_plugins_test.go
@@ -68,3 +68,24 @@ func (s *DockerSuite) TestPluginInstallImage(c *check.C) {
 	c.Assert(err, checker.NotNil)
 	c.Assert(out, checker.Contains, "content is not a plugin")
 }
+
+func (s *DockerSuite) TestPluginEnableDisableNegative(c *check.C) {
+	testRequires(c, DaemonIsLinux, ExperimentalDaemon)
+	out, _, err := dockerCmdWithError("plugin", "install", "--grant-all-permissions", pName)
+	c.Assert(err, checker.IsNil)
+	c.Assert(strings.TrimSpace(out), checker.Contains, pName)
+
+	out, _, err = dockerCmdWithError("plugin", "enable", pName)
+	c.Assert(err, checker.NotNil)
+	c.Assert(strings.TrimSpace(out), checker.Contains, "already enabled")
+
+	_, _, err = dockerCmdWithError("plugin", "disable", pName)
+	c.Assert(err, checker.IsNil)
+
+	out, _, err = dockerCmdWithError("plugin", "disable", pName)
+	c.Assert(err, checker.NotNil)
+	c.Assert(strings.TrimSpace(out), checker.Contains, "already disabled")
+
+	_, _, err = dockerCmdWithError("plugin", "remove", pName)
+	c.Assert(err, checker.IsNil)
+}

--- a/plugin/manager_linux.go
+++ b/plugin/manager_linux.go
@@ -3,6 +3,7 @@
 package plugin
 
 import (
+	"fmt"
 	"os"
 	"path/filepath"
 	"syscall"
@@ -20,6 +21,9 @@ import (
 )
 
 func (pm *Manager) enable(p *plugin) error {
+	if p.P.Active {
+		return fmt.Errorf("plugin %s is already enabled", p.Name())
+	}
 	spec, err := pm.initSpec(p)
 	if err != nil {
 		return err
@@ -27,12 +31,18 @@ func (pm *Manager) enable(p *plugin) error {
 
 	p.restartManager = restartmanager.New(container.RestartPolicy{Name: "always"}, 0)
 	if err := pm.containerdClient.Create(p.P.ID, libcontainerd.Spec(*spec), libcontainerd.WithRestartManager(p.restartManager)); err != nil { // POC-only
+		if err := p.restartManager.Cancel(); err != nil {
+			logrus.Errorf("enable: restartManager.Cancel failed due to %v", err)
+		}
 		return err
 	}
 
 	socket := p.P.Manifest.Interface.Socket
 	p.client, err = plugins.NewClient("unix://"+filepath.Join(p.runtimeSourcePath, socket), nil)
 	if err != nil {
+		if err := p.restartManager.Cancel(); err != nil {
+			logrus.Errorf("enable: restartManager.Cancel failed due to %v", err)
+		}
 		return err
 	}
 
@@ -114,6 +124,9 @@ func (pm *Manager) initSpec(p *plugin) (*specs.Spec, error) {
 }
 
 func (pm *Manager) disable(p *plugin) error {
+	if !p.P.Active {
+		return fmt.Errorf("plugin %s is already disabled", p.Name())
+	}
 	if err := p.restartManager.Cancel(); err != nil {
 		logrus.Error(err)
 	}

